spring-transaction/src/test/kotlin/org/jetbrains/exposed/spring/SpringTransactionManagerTest.kt

```diff
@@ -0,0 +1,341 @@
+    }
+
+    @Test
+    fun `transaction commit with transaction aware data source proxy`() {
+        val transactionAwareDs = TransactionAwareDataSourceProxy(ds1)
+        val tm = SpringTransactionManager(transactionAwareDs)
+        tm.executeAssert()
+
+        assertTrue(con1.verifyCallOrder("setReadOnly", "setAutoCommit", "commit"))
+        assertEquals(1, con1.commitCallCount)
+        assertTrue(con1.closeCallCount > 0)
+    }
+
+    @Test
+    fun `transaction rollback with transaction aware data source proxy`() {
+        val transactionAwareDs = TransactionAwareDataSourceProxy(ds1)
+        val tm = SpringTransactionManager(transactionAwareDs)
+        val ex = RuntimeException("Application exception")
+        try {
+            tm.executeAssert {
+                throw ex
+            }
+        } catch (e: Exception) {
+            assertEquals(ex, e)
+        }
+
+        assertTrue(con1.verifyCallOrder("setReadOnly", "setAutoCommit", "rollback"))
+        assertEquals(1, con1.rollbackCallCount)
+        assertTrue(con1.closeCallCount > 0)
+    }
+
+    @Test
+    fun `transaction exception on commit and rollback on commit failure`() {
+        con1.mockCommit = { throw SQLException("Commit failure") }
+
+        val tm = SpringTransactionManager(ds1)
+        tm.isRollbackOnCommitFailure = true
+        assertFailsWith<TransactionSystemException> {
+            tm.executeAssert()
+        }
+
+        assertTrue(con1.verifyCallOrder("setReadOnly", "setAutoCommit", "commit", "isClosed", "rollback", "close"))
+        assertEquals(1, con1.commitCallCount)
+        assertEquals(1, con1.rollbackCallCount)
+        assertEquals(1, con1.closeCallCount)
+    }
+
+    @Test
+    fun `transaction with exception on rollback`() {
+        con1.mockRollback = { throw SQLException("Rollback failure") }
+
+        val tm = SpringTransactionManager(ds1)
+        assertFailsWith<TransactionSystemException> {
+            tm.executeAssert {
+                assertEquals(false, it.isRollbackOnly)
+                it.setRollbackOnly()
+                assertEquals(true, it.isRollbackOnly)
+            }
+        }
+
+        assertTrue(con1.verifyCallOrder("setReadOnly", "setAutoCommit", "isClosed", "rollback", "close"))
+        assertEquals(1, con1.rollbackCallCount)
+        assertEquals(1, con1.closeCallCount)
+    }
+
+    private fun SpringTransactionManager.executeAssert(
+        initializeConnection: Boolean = true,
+        body: (TransactionStatus) -> Unit = {}
+    ) {
+        TransactionTemplate(this).executeWithoutResult {
+            assertEquals(
+                TransactionManager.managerFor(TransactionManager.currentOrNull()?.db),
+                TransactionManager.manager
+            )
+            if (initializeConnection) TransactionManager.current().connection
+            body(it)
+        }
+    }
+
+    private open class DataSourceSpy(private val connectionSpy: (Connection) -> Connection) : DataSource {
```

Comment:
please consider extracting utility classes to a separate with internal modifier

Reply:
@e5l Just to clarify your question

1. move the spy classes outside of the test class and then change them to internal classes, 2. just change them to internal classes, or 3. move them to a new file?

Reply:
a new file is preferred

Reply:
I created a file called TestSpys.kt that contains a collection of spy classes.

I think it's a bit of a bad name, but I'll change it if you have any suggestions.

Reply:
I would name it same as a class by default

Reply:
@FullOfOrange Just to clarify above, feel free to create 2 new files, each with its own internal class, named after the class. So 1 for `DataSourceSpy` and 1 for `ConnectionSpy`.